File: yukicoder/tyama_yukicoder187.py
#!/usr/bin/python
from functools import reduce
import sys
if sys.version_info[0]>=3: raw_input=input

def egcd(x,y):
	if y==0: return (x,1,0)
	g,a,b=egcd(y,x%y)
	return (g,b,a-x//y*b)

def perform(a1,m1,a2,m2):
	g,x,y=egcd(m1,m2)
	if (a2-a1)%g: raise ValueError
	l=m1//g*m2
	return ((a1+(a2-a1)//g*x*m1)%l,l)

def chinese(a):
	if not a: return None
	try:
		return reduce(lambda x,y:perform(*(x+y)),a)
	except ValueError:
		return None

a=[]
for _ in range(int(raw_input())):
	a.append(tuple(int(e) for e in raw_input().split()))
r=chinese(a)
print(r[not r[0]]%(10**9+7) if r else -1)

File: yukicoder/tyama_yukicoder187.rb
#!/usr/bin/ruby
def egcd(x,y)
	return [x,1,0] if y==0
	g,a,b=egcd(y,x%y)
	[g,b,a-x/y*b]
end
def chinese(a)
	return nil if !a||a.empty?
	a.reduce{|(a1,m1),(a2,m2)|
		g,x,y=egcd(m1,m2)
		break if (a2-a1)%g>0
		l=m1/g*m2
		[(a1+(a2-a1)/g*x*m1)%l,l]
	}
end
r=chinese($<.drop(1).map{|e|e.split.map &:to_i})
p r ?r[r[0]==0?1:0]%1000000007:-1

__END__
f=->x,y{return[x,1,0]if y==0
g,a,b=f[y,x%y]
[g,b,a-x/y*b]}
gets
a=$<.map{|e|e.split.map &:to_i}
p (r=a.reduce{|(a,m),(b,n)|g,x,y=f[m,n]
(b-=a)%g>0&&break
n*=m/g
[(a+b/g*x*m)%n,n]})?r[r[0]>0?0:1]%(10**9+7):-1
#[(a+b/g*x*m)%n,n]})?r[~(0<=>r[0])]%(10**9+7):-1
